browser/src/statechart.js

```diff
@@ -0,0 +1,615 @@
+    // tools not available in RGB mode
+    'keydown.t': {
+      target: '.threshold',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+    'keydown.m': {
+      target: '.autofit',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+    'keydown.w': {
+      target: '.watershed',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+/**
+ * Handles tool behavior with a loading state.
+ */
+const mouseState = {
+  initial: 'toolbar',
+  states: {
+    toolbar: toolbarState,
+    loading: {
+      on: { LOADED: 'toolbar.hist' }
+    }
+  },
+  on: {
+    EDIT: { actions: forwardTo('backend') },
+    UNDO: { actions: forwardTo('backend') },
+    REDO: { actions: forwardTo('backend') },
+  }
+};
+
+/**
+ * Handles adjustments to the images displayed on the canvas.
+ */
+const adjusterState = {
+  on: {
+    SCROLLCONTRAST: { actions: 'changeContrast' },
+    SCROLLBRIGHTNESS: { actions: 'changeBrightness' },
+    'keydown.h': { actions: 'toggleHighlight' },
+    'keydown.0': { actions: 'resetBrightnessContrast' },
+    'keydown.i': { actions: 'toggleInvert' }
+  }
+};
+
+/**
+ * Handles moving around the canvas.
+ */
+const canvasState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: { 'keydown.space': 'pan' },
+    },
+    pan: {
+      on: {
+        'keyup.space': 'idle',
+        mousemove: { actions: ['pan', 'updateMousePos'] }
+      }
+    }
+  },
+  on: {
+    'keydown.-': { actions: send({ type: 'ZOOM', change: -1 }) },
+    'keydown.=': { actions: send({ type: 'ZOOM', change: 1 }) },
+    ZOOM: { actions: 'zoom' },
+    mousemove: { actions: 'updateMousePos' }
+  }
+};
+
+const brushState = {
+  on: {
+    'keydown.up': { actions: 'incrementBrushSize' },
+    'keydown.down': { actions: 'decrementBrushSize' }
+  }
+};
+
+const displayState = {
+  initial: 'overlay',
+  states: {
+    overlay: {
+      on: {
+        'keydown.z': {
+          target: 'labels',
+          actions: 'updateCanvas'
+        }
+      }
+    },
+    labels: {
+      on: {
+        'keydown.z': {
+          target: 'raw',
+          actions: 'updateCanvas'
+        }
+      }
+    },
+    raw: {
+      on: {
+        'keydown.z': {
+          target: 'overlay',
+          actions: 'updateCanvas'
+        }
+      }
+    }
+  }
+}
+
+const frameState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        'keydown.a': { actions: 'decrementFrame' },
+        'keydown.left': { actions: 'decrementFrame' },
+        'keydown.d': { actions: 'incrementFrame' },
+        'keydown.right': { actions: 'incrementFrame' },
+        'keydown.c': { actions: 'incrementChannel' },
+        'keydown.C': { actions: 'decrementChannel' },
+        'keydown.f': { actions: 'incrementFeature' },
+        'keydown.F': { actions: 'decrementFeature' },
+        LOADING: 'loading',
+      }
+    },
+    loading: {
+      on: { LOADED: 'idle' }
+    }
+  },
+  on: {
+    SETFRAME: { actions: forwardTo('backend') },
+  }
+};
+
+const confirmState = {
```

Comment:
<img width="830" alt="Screen Shot 2021-02-16 at 2 17 04 PM" src="https://user-images.githubusercontent.com/40068392/108131938-e96ad580-7066-11eb-9794-72b3a64028f7.png">

The confirm state manages keybind actions. All of these actions require confirmation with Enter before launching them, or can be aborted with Escape. The view displays a prompt depending on which of the confirm states we are in. This state also has a loading substate so we cannot start an action while waiting for a server response.

Reply:
So to clarify, the keybinds for both `confirmState` and `frameState` are globally available, but those in the `confirmState` ask for Enter/Esc press and `frameState` does not?

Reply:
The frameState does not edit the labels, it just fetches different image slices from the backend, so there's no need to confirm those actions. The actions in confirm state do edit labels, so we ask the user to confirm the change in case they hit the keybind accidentally.